donkeycar/pipeline/sequence.py

```diff
@@ -1,154 +1,90 @@
-        return record
-
-    next = __next__
+    def __iter__(self) -> Iterator[X]:
+        pass
 
 
-class TfmIterator(Generic[R, XOut, YOut],  SizedIterator[Tuple[XOut, YOut]]):
+class _BaseTfmIterator(Generic[X, Y, XOut, YOut], Iterator[Tuple[XOut, YOut]]):
+    """
+    A basic transforming iterator. Do no use this class directly.
+    """
     def __init__(self,
-                 iterable: Iterable[R],
-                 x_transform: Callable[[R], XOut],
-                 y_transform: Callable[[R], YOut]) -> None:
-
+                 iterable: 'TfmShallowList') -> None:
         self.iterable = iterable
-        self.x_transform = x_transform
-        self.y_transform = y_transform
-        self.iterator = BaseTfmIterator_(
-            iterable=self.iterable, x_transform=self.x_transform, y_transform=self.y_transform)
+        self.iterator: Iterator[Union[TubRecord, Tuple[X, Y]]] \
+            = iter(self.iterable.iterable)
 
-    def __len__(self):
-        return len(self.iterator)
+    def __iter__(self) -> '_BaseTfmIterator[X, Y, XOut, YOut]':
+        return self
 
-    def __iter__(self) -> SizedIterator[Tuple[XOut, YOut]]:
-        return BaseTfmIterator_(
-            iterable=self.iterable, x_transform=self.x_transform, y_transform=self.y_transform)
+    def __next__(self) -> Tuple[XOut, YOut]:
+        r: Union[TubRecord, Tuple[X, Y]] = next(self.iterator)
+        return self.iterable.transform_item(r)
 
-    def __next__(self):
-        return next(self.iterator)
 
-
-class TfmTupleIterator(Generic[X, Y, XOut, YOut],  SizedIterator[Tuple[XOut, YOut]]):
+class TfmShallowList(Generic[XOut, YOut], SizedIterable[Tuple[XOut, YOut]]):
     def __init__(self,
-                 iterable: Iterable[Tuple[X, Y]],
-                 x_transform: Callable[[X], XOut],
-                 y_transform: Callable[[Y], YOut]) -> None:
-
+                 iterable: Union[List[TubRecord],
+                                 'TfmShallowList[XOut, YOut]'],
+                 x_transform: Callable[[Any], XOut],
+                 y_transform: Callable[[Any], YOut]) -> None:
         self.iterable = iterable
         self.x_transform = x_transform
         self.y_transform = y_transform
-        self.iterator = BaseTfmIterator_(
-            iterable=self.iterable, x_transform=self.x_transform, y_transform=self.y_transform)
-
-    def __len__(self):
-        return len(self.iterator)
 
-    def __iter__(self) -> SizedIterator[Tuple[XOut, YOut]]:
-        return BaseTfmIterator_(
-            iterable=self.iterable, x_transform=self.x_transform, y_transform=self.y_transform)
+    def __len__(self) -> int:
+        return len(self.iterable)
 
-    def __next__(self):
-        return next(self.iterator)
+    def __iter__(self) -> _BaseTfmIterator[X, Y, XOut, YOut]:
+        return _BaseTfmIterator(self)
 
+    def transform_item(self, r: Union[TubRecord, Tuple[X, Y]]) \
+            -> Tuple[XOut, YOut]:
+        if isinstance(r, tuple) and len(r) == 2:
```

Comment:
Here is were all the magic happens, the other objects and code are only protocol implementations.